legend-engine-executionPlan-execution-store-relational-connection/src/main/java/org/finos/legend/engine/plan/execution/stores/relational/connection/authentication/strategy/SnowflakePublicAuthenticationStrategy.java

```diff
@@ -0,0 +1,111 @@
+import org.finos.legend.engine.plan.execution.stores.relational.connection.driver.DatabaseManager;
+import org.finos.legend.engine.plan.execution.stores.relational.connection.ds.DataSourceWithStatistics;
+import org.finos.legend.engine.shared.core.vault.Vault;
+import org.pac4j.core.profile.CommonProfile;
+
+import javax.security.auth.Subject;
+import java.io.StringReader;
+import java.security.PrivateKey;
+import java.security.Security;
+import java.sql.Connection;
+import java.sql.SQLException;
+import java.util.Properties;
+
+public class SnowflakePublicAuthenticationStrategy extends AuthenticationStrategy
+{
+    private final String privateKeyVaultReference;
+    private final String passPhraseVaultReference;
+    private final String publicUserName;
+
+    public SnowflakePublicAuthenticationStrategy(String privateKeyVaultReference, String passPhraseVaultReference, String publicUserName)
+    {
+        this.privateKeyVaultReference = privateKeyVaultReference;
+        this.passPhraseVaultReference = passPhraseVaultReference;
+        this.publicUserName = publicUserName;
+    }
+
+    @Override
+    public String getLogin()
+    {
+        return this.publicUserName;
+    }
+
+    @Override
+    public String getPassword()
+    {
+        return null;
+    }
+
+    @Override
+    public Pair<String, Properties> handleConnection(String url, Properties properties, DatabaseManager databaseManager)
+    {
+        Properties connectionProperties = new Properties();
+        connectionProperties.putAll(properties);
+        connectionProperties.put("privateKey", getEncryptedPrivateKey(this.privateKeyVaultReference, this.passPhraseVaultReference));
+        connectionProperties.put("user", this.publicUserName);
+        return Tuples.pair(url, connectionProperties);
+    }
+
+    @Override
+    protected Connection getConnectionImpl(DataSourceWithStatistics ds, Subject subject, MutableList<CommonProfile> profiles) throws ConnectionException
+    {
+        try
+        {
+            return ds.getDataSource().getConnection();
+        }
+        catch (SQLException e)
+        {
+            throw new ConnectionException(e);
+        }
+    }
+
+    @Override
+    public AuthenticationStrategyKey getKey()
+    {
+        return new SnowflakePublicAuthenticationStrategyKey(this.privateKeyVaultReference, this.passPhraseVaultReference, this.publicUserName);
+    }
+
+    private PrivateKey getEncryptedPrivateKey(String privateKeyVaultReference, String passPhraseVaultReference)
+    {
+        try
+        {
+            PrivateKeyInfo privateKeyInfo = null;
+            Security.addProvider(new BouncyCastleProvider());
+            PEMParser pemParser = new PEMParser(new StringReader(Vault.INSTANCE.getValue(privateKeyVaultReference)));
+            Object pemObject = pemParser.readObject();
+            if (pemObject instanceof PKCS8EncryptedPrivateKeyInfo)
```

Comment:
Minor.  Since this strategy only supports encrypted private keys, might be better to throw when pemObject is not a PKCS8EncryptedPrivateKeyInfo.

Reply:
I'm going to throw... I actually also have to account for other formats within this use case.